rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,132 @@
+# Integrate `cosign` with `kpack`
+
+## Problem
+
+The signing solution currently supported by `kpack` is
+[Notary v1](https://github.com/theupdateframework/notary). However, Notary
+presents itself with several different challenges, including:
+- It is unable to sign container image digests. The user must always specify a
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures.
+  - Currently, there is no built-in command for relocation on `cosign`, but
+    copying container images and signatures across registries using other
+    container CLI tools such as `crane` and then verifying with `cosign` is
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
+  configure image signing with `cosign` as the signing mechanism, specifying
+  the secret reference to get the private key and passphrase. The user can also
+  specify signature annotations as part of configuration:
+```yaml
+cosign:
+  # a list of secret references holding the private keys and passphrases for signing images
+  secretRefs:
+  - name: <secret-name 1>
+  ...
+  - name: <secret-name n>
+  annotations: # optional map of key-value pairs the user wants to add to images
```

Comment:
would it make more sense for annotations to be configured outside of the image spec such as in a ConfigMap?

1. Allows shared annotations per namespace
2. It might be easier to update a configmap than this field for non-static annotations (ex. git commit)
  - kpack cli doesn't currently support configuring the cosign field

Reply:
I believe it would be more flexible. In this case, we could have a list of `ConfigMap`s under the `annotations` key.

About `kpack` CLI not currently supporting the `cosign` field, can we create a separate RFC to address this concern?

Reply:
How commonly do we think annotations will be used, especially non-static ones?

Reply:
What if we support multiple levels of annotations?
1. Per image - highest priority & overrides any key from a lower priority
2. Per namespace - lower priority

We could implement 1 first, and add 2 later, if it's needed?

How would you imagine the workflow for annotating the git commit?

Reply:
> How commonly do we think annotations will be used, especially non-static ones?

I'd probably want to start with some automatically-created dynamic ones, e.g. the git commit/url for the source for the build, information about the build (number, trigger(s), etc.).

I'm not sure about what you might put for static ones?

Reply:
Could the prior suggestion of having the `kpack` build number as an annotation replace the requirement of having git commits as annotations, for instance?

Reply:
> I'd probably want to start with some automatically-created dynamic ones, e.g. the git commit/url for the source for the build, information about the build (number, trigger(s), etc.).

We could pretty easily do this without requiring the user to specify but, is are these git source annotations necessary? That metadata is already on image labels and the image digest (which includes image labels) will be signed by cosign.

Reply:
> Per namespace - lower priority
Is it important to specify how this might be done now at the rfc stage?

Reply:
> We could pretty easily do this without requiring the user to specify but, are these git source annotations necessary? That metadata is already on image labels and the image digest (which includes image labels) will be signed by cosign.

Good point. Maybe we don't need any auto-generated annotations?

> Is it important to specify how this might be done now at the rfc stage?

This feels like the sort of thing to spec out now.... however, I wonder if it makes sense to start with a single thing: only support annotations at the image level? Then we don't have to talk about priorities, and we don't have to talk about global vs local annotations. WDYT?

Reply:
> I'm not sure about what you might put for static ones?

Our thought was that users could want to put annotations with some static information such as a container image name, but it's possible that this is already addressed by other metadata pieces (such as build number).

Reply:
More arguments why static annotations are useful:

I pictured the signature annotations to be used help an attestation system or as a part of a end users supply chain processes.

1. Someone could put a signature on a container image at build time with an annotation saying, "Built on my build system".
2. Later on they could run integration tests on the container image saying, "Passed integration tests".  Let's say both those processes occur at VMware and they would have the same signing key.  If I remember correctly, you can only sign an image with Cosign more than once if the two signatures have different annotations.
3. Now, let's say the end user has a policy that they have to bless every image that comes into their air-gapped environment.  After image relocation (which would preserve the original keys and their annotations) and some pass through a test environment they would put another signature, with another annotation that says "this image is allowed to be promoted to production".

Now, maybe only the first stage in this example would actually be signed with kpack, but I would like enable the rest of this chain of signatures by using kpack to allow its users to apply signatures with static annotations.

Reply:
>Someone could put a signature on a container image at build time with an annotation saying, "Built on my build system".

Would it be safer to rely on secure identity built into the systems themselves (@anvega can hopefully provide more detail here) rather than a user-supplied value for a build server's identity?

Also cc @joshuagl in case you've got thoughts 😄

As I wrote in another comment thread, given that signature annotations are a thing, it seems reasonable to support them at the Image level, unless folks have significant reasons not to.

Reply:
cosign author here! I put the "annotations" into the tool to basically act as "lightweight" attestations. There are some other richer formats we're looking to support down the road (mostly the in-toto Attestation format: https://github.com/in-toto/attestation) but nothing is concrete yet.

I hope that helps!

Reply:
Aside from the constraint pointed out of annotations meant to support lightweight attestations, secure identities are inherently short-lived. As such, there'd be limitations around how to validate those without additional moving parts or a broader framework. 

That said, consideration for secure identities is directionally correct. With support for in-toto down the line, it does open up the possibilities to integrate in-toto attestations with secure identities.